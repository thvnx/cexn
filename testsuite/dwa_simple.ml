(* This file is part of mlexn.

   mlexn is free software: you can redistribute it and/or modify it under the
   terms of the GNU General Public License as published by the Free Software
   Foundation, either version 3 of the License, or (at your option) any later
   version.

   mlexn is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
   FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
   details.

   You should have received a copy of the GNU General Public License along with
   mlexn.  If not, see <http://www.gnu.org/licenses/>. *)

let dwa x =
  Printf.printf "add_float      %s\n" (Dwa.to_string (Dwa.add_float (Dwa.of_float x) x));
  Printf.printf "sloppy_add     %s\n" (Dwa.to_string (Dwa.sloppy_add (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "add            %s\n" (Dwa.to_string (Dwa.add (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "sub            %s\n" (Dwa.to_string (Dwa.sub (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "mul_float      %s\n" (Dwa.to_string (Dwa.mul_float (Dwa.of_float x) x));
  Printf.printf "mul_float_fast %s\n" (Dwa.to_string (Dwa.mul_float_fast (Dwa.of_float x) x));
  Printf.printf "mul_float_fma  %s\n" (Dwa.to_string (Dwa.mul_float_fma (Dwa.of_float x) x));
  Printf.printf "mul            %s\n" (Dwa.to_string (Dwa.mul (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "mul_fma1       %s\n" (Dwa.to_string (Dwa.mul_fma1 (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "mul_fma        %s\n" (Dwa.to_string (Dwa.mul_fma (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "div_float1     %s\n" (Dwa.to_string (Dwa.div_float1 (Dwa.of_float x) x));
  Printf.printf "div_float2     %s\n" (Dwa.to_string (Dwa.div_float2 (Dwa.of_float x) x));
  Printf.printf "div_float      %s\n" (Dwa.to_string (Dwa.div_float (Dwa.of_float x) x));
  Printf.printf "div1           %s\n" (Dwa.to_string (Dwa.div1 (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "div            %s\n" (Dwa.to_string (Dwa.div (Dwa.of_float x) (Dwa.of_float x)));
  Printf.printf "div_fma        %s\n" (Dwa.to_string (Dwa.div_fma (Dwa.of_float x) (Dwa.of_float x)))

let _ =
  List.iter (fun x ->
      Printf.printf "#DWA tests, x = %h\n" x;
      dwa x;
      Printf.printf "\n") [1.; 2.; 4.; 8.; ~-. 1.; ~-. 2.; ~-. 4.; ~-. 8.]
