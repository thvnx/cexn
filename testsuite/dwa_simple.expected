#DWA tests, x = 0x1p+0
add_float      0x1p+1 0x0p+0
sloppy_add     0x1p+1 0x0p+0
add            0x1p+1 0x0p+0
mul_float      0x1p+0 0x0p+0
mul_float_fast 0x1p+0 0x0p+0
mul_float_fma  0x1p+0 0x0p+0
mul            0x1p+0 0x0p+0
mul_fma1       0x1p+0 0x0p+0
mul_fma        0x1p+0 0x0p+0
div_float1     0x1p+0 0x0p+0
div_float2     0x1p+0 0x0p+0
div_float      0x1p+0 0x0p+0
div1           0x1p+0 0x0p+0
div            0x1p+0 0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = 0x1p+1
add_float      0x1p+2 0x0p+0
sloppy_add     0x1p+2 0x0p+0
add            0x1p+2 0x0p+0
mul_float      0x1p+2 0x0p+0
mul_float_fast 0x1p+2 0x0p+0
mul_float_fma  0x1p+2 0x0p+0
mul            0x1p+2 0x0p+0
mul_fma1       0x1p+2 0x0p+0
mul_fma        0x1p+2 0x0p+0
div_float1     0x1p+0 0x0p+0
div_float2     0x1p+0 0x0p+0
div_float      0x1p+0 0x0p+0
div1           0x1p+0 0x0p+0
div            0x1p+0 0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = 0x1p+2
add_float      0x1p+3 0x0p+0
sloppy_add     0x1p+3 0x0p+0
add            0x1p+3 0x0p+0
mul_float      0x1p+4 0x0p+0
mul_float_fast 0x1p+4 0x0p+0
mul_float_fma  0x1p+4 0x0p+0
mul            0x1p+4 0x0p+0
mul_fma1       0x1p+4 0x0p+0
mul_fma        0x1p+4 0x0p+0
div_float1     0x1p+0 0x0p+0
div_float2     0x1p+0 0x0p+0
div_float      0x1p+0 0x0p+0
div1           0x1p+0 0x0p+0
div            0x1p+0 0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = 0x1p+3
add_float      0x1p+4 0x0p+0
sloppy_add     0x1p+4 0x0p+0
add            0x1p+4 0x0p+0
mul_float      0x1p+6 0x0p+0
mul_float_fast 0x1p+6 0x0p+0
mul_float_fma  0x1p+6 0x0p+0
mul            0x1p+6 0x0p+0
mul_fma1       0x1p+6 0x0p+0
mul_fma        0x1p+6 0x0p+0
div_float1     0x1p+0 0x0p+0
div_float2     0x1p+0 0x0p+0
div_float      0x1p+0 0x0p+0
div1           0x1p+0 0x0p+0
div            0x1p+0 0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = -0x1p+0
add_float      -0x1p+1 0x0p+0
sloppy_add     -0x1p+1 0x0p+0
add            -0x1p+1 0x0p+0
mul_float      0x1p+0 0x0p+0
mul_float_fast 0x1p+0 0x0p+0
mul_float_fma  0x1p+0 0x0p+0
mul            0x1p+0 0x0p+0
mul_fma1       0x1p+0 0x0p+0
mul_fma        0x1p+0 0x0p+0
div_float1     0x1p+0 -0x0p+0
div_float2     0x1p+0 -0x0p+0
div_float      0x1p+0 -0x0p+0
div1           0x1p+0 -0x0p+0
div            0x1p+0 -0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = -0x1p+1
add_float      -0x1p+2 0x0p+0
sloppy_add     -0x1p+2 0x0p+0
add            -0x1p+2 0x0p+0
mul_float      0x1p+2 0x0p+0
mul_float_fast 0x1p+2 0x0p+0
mul_float_fma  0x1p+2 0x0p+0
mul            0x1p+2 0x0p+0
mul_fma1       0x1p+2 0x0p+0
mul_fma        0x1p+2 0x0p+0
div_float1     0x1p+0 -0x0p+0
div_float2     0x1p+0 -0x0p+0
div_float      0x1p+0 -0x0p+0
div1           0x1p+0 -0x0p+0
div            0x1p+0 -0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = -0x1p+2
add_float      -0x1p+3 0x0p+0
sloppy_add     -0x1p+3 0x0p+0
add            -0x1p+3 0x0p+0
mul_float      0x1p+4 0x0p+0
mul_float_fast 0x1p+4 0x0p+0
mul_float_fma  0x1p+4 0x0p+0
mul            0x1p+4 0x0p+0
mul_fma1       0x1p+4 0x0p+0
mul_fma        0x1p+4 0x0p+0
div_float1     0x1p+0 -0x0p+0
div_float2     0x1p+0 -0x0p+0
div_float      0x1p+0 -0x0p+0
div1           0x1p+0 -0x0p+0
div            0x1p+0 -0x0p+0
div_fma        0x1p+0 0x0p+0

#DWA tests, x = -0x1p+3
add_float      -0x1p+4 0x0p+0
sloppy_add     -0x1p+4 0x0p+0
add            -0x1p+4 0x0p+0
mul_float      0x1p+6 0x0p+0
mul_float_fast 0x1p+6 0x0p+0
mul_float_fma  0x1p+6 0x0p+0
mul            0x1p+6 0x0p+0
mul_fma1       0x1p+6 0x0p+0
mul_fma        0x1p+6 0x0p+0
div_float1     0x1p+0 -0x0p+0
div_float2     0x1p+0 -0x0p+0
div_float      0x1p+0 -0x0p+0
div1           0x1p+0 -0x0p+0
div            0x1p+0 -0x0p+0
div_fma        0x1p+0 0x0p+0

